Add task models

pcreech opened the pull request:
Add task models.

The tasking system is a pretty solid, well baked solution, so making too many changes at this time might not make sense.  I tried to make ones that made sense in relation to utilizing a relational database.

Utilized foreign keys where it appears to make sense, and converted some of the data types to utilize their natural data state in the database (mainly DateTimes)

There are four fields that will need more thought, but wanted to go ahead and put up the PR to start the conversation. These fields store complex data structures, that could potentially vary based on utilization.  

These are:
- Task.result
- Task.error
- ScheduledCalls.args
- ScheduledCalls.kwargs

My initial thought was to convert them into properties, so we can split up the internal representation (TextField) with an external implementation of some type of complex object.  This should be transparent for the user, as when they use "scheduled_call.args" it should execute the code in the property function.  Currently the implementation is a naive example implementation utilizing json.dumps, that will need to be fleshed out.  

With that, I'm also looking into how creating a custom Django field (Michael's suggestion) could provide the same functionality.  It appears to provide a more natural interface for converting between database state and python state.  Brian suggested utilizing django.picklefield.  This is a viable and enticing option, but would become another dependency we would have to manage.  We should probably weigh the cost of developing our own solution with the cost of introducing another dependency.

closes [#2087](https://pulp.plan.io/issues/2087)

pcreech commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,296 @@
+    _error = models.TextField(null=True, db_column='error')
+    _result = models.TextField(null=True, db_column='result')
+
+    parent = models.ForeignKey("Task", null=True)
```
I converted spwaned_tasks: [] into a parent foreign key.  The other option is to create a table with task ids.  This was based on the assumption that I came to that the new task.id uuid is the same as the uuid's describing tasks in spanwed_tasks array.

pcreech commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,296 @@
+    lock = models.TextField(unique=True, null=False, choices=LOCK_STRINGS)
+
+
+class Task(Model):
```
I removed task_id from the model, with the assumption that the new uuid pk will act as the task_id.

mhrivnak replied:
I think we still want every user-facing model to have a "natural key" that is different from the pk created by django. @seandst I recall you being enthusiastic about that idea- can you comment on this case?

seandst replied:
In this case I think I'm not seeing anything on the model that would be useful in creating a natural key, since `task_type` combined with `task_group` is definitely not going to be unique, and there are enough potential problems (time collisions, mainly) using the `started_at` stamp that it's probably best to leave it out of the natural key. I don't think that leaves enough for constructing a proper natural key.

pcreech commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,296 @@
+
+    name = models.TextField(db_index=True, unique=True)
+    timestamp = models.DateTimeField(auto_now_add=True)
+    lock = models.TextField(unique=True, null=False, choices=LOCK_STRINGS)
```
I consolidated CeleryBeatLock and ResourceManagerLock into a single TaskLock, since the data structures were near identical, and utilized a TextField for the lock.  Instead of both using the string 'locked' they now use a context specific string to identify the lock.

mhrivnak replied:
+1 I've been thinking the same thing for a while.

pcreech commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,296 @@
+    resource = models.TextField()
+
+    task = models.ForeignKey(Task, unique=True)
+    worker = models.ForeignKey(Worker)
```
Instead of utilizing strings for the task_id and worker_name, I converted these fields into foreign keys to tie them to the task and worker.  This is in line with converting to a relational model, but might need to be revisited if it causes headaches for the reserved resource code.

pcreech commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,296 @@
+
+    last_updated = models.DateTimeField(auto_now=True)
+
+    principal = models.ForeignKey("User", null=True)
```
This was a stored user object.  I converted this to be a foreign key on the currently non-existent "User" table.  There was question of if this is still needed/utilized.

mhrivnak replied:
It wouldn't hurt to check code paths, but I'm pretty sure this isn't used. We don't track which user created any other objects in pulp. It's a reasonable idea to consider, but I think we should re-consider that from a global perspective later. I suggest dropping it from this model.

pcreech commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,296 @@
+
+    task = models.ForeignKey(Task, related_name="tags", related_query_name="tag")
+
+class ScheduledCalls(Model):
```
This model was never converted to mongoengine, and there were statements that the pymongo object had values that no longer were used.  This model should probably be scrutinized heavily to ensure data wasn't missed.  I did a best guess between what's in the pymongo object and what is stored in the database when a schedule is created.

mhrivnak commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,296 @@
+
+    task = models.ForeignKey(Task, unique=True)
+    worker = models.ForeignKey(Worker)
+
```
2 lines for pep8

mhrivnak commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,296 @@
+    """
+    name = models.TextField(db_index=True, unique=True)
+    last_heartbeat = models.DateTimeField()
+
```
2 lines for pep8

mhrivnak commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,296 @@
+    name = models.TextField()
+
+    task = models.ForeignKey(Task, related_name="tags", related_query_name="tag")
+
```
2 lines for pep8

mhrivnak commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,296 @@
+
+    """
+    task = models.TextField()
+    enabled = models.BooleanField()
```
I suggest a default of ``True`` since it will be the very common case.

mhrivnak commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,296 @@
+    enabled = models.BooleanField()
+    resource = models.TextField(null=True)
+
+    consecutive_failures = models.IntegerField()
```
We should remove both of these fields and remove the feature of auto-disabling scheduled tasks. It adds a lot of complexity without much value.

I added a task just now to track that. Feedback welcome: https://pulp.plan.io/issues/2176

mhrivnak commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,296 @@
+    :cvar last_run_at: Last time this schedule was ran
+    :type last_run_at: models.DateTimeField
+
+    :cvar last_updated: The last time this schedule was saved to the database
```
I think this is specifically the last time a user changed something about the schedule. This field helps the ``Scheduler`` know if it needs to re-load schedules from the database when considering whether any are due to be run. I do not think this field should be touched when something like the ``total_run_count`` is incremented.

mhrivnak commented:
Don't forget to cross-link the PR with the redmine issue so people can easily find each from the other.

https://pulp.plan.io/issues/2087

mhrivnak commented:
@bmbouter should definitely give this a thorough review.

Once the model is agreed on, you'll of course still need to add the docs about how to map old data to new.

mhrivnak commented:
Speaking of mapping old data to new, I propose that we document that of these models, only the scheduled call will have data migrated from mongo to postgres. Most of the others should be empty anyway. Task status is something that gets cleaned up on a regular basis, so it's understood to not be indefinitely persistent. I think leaving that history behind should be fine, and will give us the flexibility to model it however is best in 3.0.

jortel commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,298 @@
+
+    Fields:
+
+    :cvar task_type: The type of task this is
```
The task_ here is redundant.

jortel commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,298 @@
+    :cvar task_type: The type of task this is
+    :type task_type: models.TextField
+
+    :cvar task_group: The group this task belongs to
```
The `task_` here is redundant.  Just `group` is sufficient.

jortel commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,298 @@
+        (CANCELED, 'Canceled')
+    )
+    task_type = models.TextField()
+    task_group = models.UUIDField()
```
Same `task_` comment as above.

jortel commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,298 @@
+    finish_time = models.DateTimeField()
+
+    _error = models.TextField(null=True, db_column='error')
+    _result = models.TextField(null=True, db_column='result')
```
I think we should avoid fields with leading `_`.  What is the purpose of this?

pcreech replied:
The purpose here is to use python's way of marking a variable 'private', i.e. don't use this.  In the naive implementation I did to get the PR up, the choice I made for abstracting the db implementation from the user is to utilize python properties to expose the functionality I wanted.  When using "task.result" the property below will intercept the variable call, and convert it for general consumption from the internal storage mechanism to the public interaction.

This is a common thing to do in .Net with properties, and it seemed to hold true for python as well.

That being said, so far my investigation is leaning me towards converting these into some type of PickleField type, whether we use django.picklefield or implement our own.

jortel commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,298 @@
+    worker = models.ForeignKey(Worker)
+
+    @property
+    def result(self):
```
I suggest instead of using `_error` field and a _property_ to deal with JSON, we use a custom field (if not already exists) that does this.

jortel replied:
Perhaps: https://docs.djangoproject.com/en/dev/ref/contrib/postgres/fields/#django.contrib.postgres.fields.JSONField

mhrivnak replied:
I suggested the same. I think the note on the PR says he's still investigating options. Thanks for pointing out that ``JSONField``!

pcreech replied:
I'm hesitant to utilize JSONField, as that locks us into very recent versions of postgres for little benefit (these fields don't get utilized in the database, they are just for persistent data storage (based on a conversation I had with @bmbouter ) 

With that, as mentioned in a previous comment, i'm looking into a PickleField implementation that will be more db-agnostic.

jortel commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,298 @@
+            return self._result_object
+
+    @property
+    def error(self):
```
I suggest instead of using `_error` field and a _property_ to deal with JSON, we use a custom field (if not already exists) that does this.

pcreech replied:
As mentioned in earlier comments, I will be looking into a pickefield implementation

jortel commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,298 @@
+        super(Model, self).save(*args, **kwargs)
+
+
+class TaskTags(Model):
```
Tables are implicitly plural and each model is a instance (row).  `TaskTag` (singular) would be more appropriate.

pcreech replied:
Great catch!

jortel commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,298 @@
+
+    def save(self, *args, **kwargs):
+        try:
+            self._args = json.dumps(self._error_object)
```
What is `self._args`?  Where does `self._error_object` come from?

pcreech replied:
copy/paste errors.  should be self._error

self._error_object gets created in the property getter, therefore if it doesn't exist, it never got utilized and never modified

jortel commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,298 @@
+
+    remaining_runs = models.IntegerField()
+    total_run_count = models.IntegerField()
+    last_run_at = models.DateTimeField()
```
Suggest dropping the `_at` and just be`last_run`.  The `_at` adds no value and better symmetry with `first_run`.

pcreech replied:
Was more in the interest here of 1-1 variable naming, facilitating hopefully minimal code refactoring to utilize this model. Can easily change if others are in agreement.

jortel commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,298 @@
+    principal = models.ForeignKey("User", null=True)
+
+    _args = models.TextField(db_column="args")
+    _kwargs = models.TextField(db_column="kwargs")
```
Same comment here about the `_` prefixed fields and properties.  Let make this simpler by using a custom field.

pcreech replied:
Agreed, as mentioned in previous comments I'm looking into a picklefield implementation

jortel commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,298 @@
+    total_run_count = models.IntegerField()
+    last_run_at = models.DateTimeField()
+
+    last_updated = models.DateTimeField()
```
Suggest:  `models.DateTimeField(auto_now=True, blank=True, null=True)`

mhrivnak replied:
Just to emphasize my note above, this field should only be updated when the user changes a property, like the schedule itself, or the arguments. It should not be update when something like the ``remaining_runs`` value gets updated.

You can see the usage here: https://github.com/pulp/pulp/blob/0128ae2c99d9d57c4f73bcde52b43e1c81891736/server/pulp/server/async/scheduler.py#L328

jortel replied:
Makes sense.

jortel commented on platform/pulp/platform/models/__init__.py:
```diff
@@ -4,3 +4,5 @@
 from .consumer import Consumer, ConsumerContent  # NOQA
 from .repository import Distributor  # NOQA
+
+from .task import (ReservedResource, Worker, Task, TaskTags, TaskLock, ScheduledCalls) #NOQA
```
Don't need the `()` since it all fits on 1 line.

jortel commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,298 @@
+    state = models.TextField(choices=STATES)
+
+    start_time = models.DateTimeField()
+    finish_time = models.DateTimeField()
```
I'm a little conflicted about the `_time` in these field names.  They're actually a `datetime` and we don't normally include the datatype in field names.  Perhaps: `started` and `finished`?  I don't feel strongly about this.

pcreech replied:
Was more in the interest here of 1-1 variable naming, facilitating hopefully minimal code refactoring to utilize this model.  Can easily change if others are in agreement.

mhrivnak replied:
I can see it both ways. Just intuitively, I can imagine someone's first impression being that ``task.started`` and ``task.finished`` could be booleans. Generally since python isn't typed, if it's not reasonably obvious what type something will be, I'm not opposed to making it clear in the variable or attribute name. But I could easily see it both ways here and would be happy with either.

pcreech commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,231 @@
+    state = models.TextField(choices=STATES)
+
+    started_at = models.DateTimeField()
+    finished_at = models.DateTimeField()
```
Changed start_time and finish_time to started_ad and finished_at.

Per @jortel 's comments, I felt this was a sane modification as time felt off, but started and finished seemed like Booleans intuitively to me.

pcreech commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,231 @@
+    finished_at = models.DateTimeField()
+
+    error = JSONField()
+    result = JSONField()
```
Removed the property implementation, and replaced it with a custom JSONField implementation

pcreech commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,231 @@
+    last_updated = models.DateTimeField()
+
+    args = JSONField()
+    kwargs = JSONField()
```
Removed the property implementation, and replaced it with a custom JSONField implementation

pcreech commented:
Updated from latest feedback.

It was decided to create a JSONField implementation instead of utilizing a Picklefield due to potential issues with pickling.

I've added a skeleton, broken JSONFIeld implementation for the moment, so comment can be made on the models/task.py items.  I'm currently actively working on the JSONField implementation.

mhrivnak commented on platform/pulp/platform/fields.py:
```diff
@@ -0,0 +1,14 @@
+from django.db import models
+
+class JSONField(models.TextField):
+    def __init__(self, *args, **kwargs):
```
Not sure you'll need to override this, but obviously this class is a WIP. Looks good.

pcreech replied:
Yeah, I just committed a skeleton version at the moment.  Definitely WIP

pcreech commented on platform/pulp/platform/fields.py:
```diff
@@ -0,0 +1,96 @@
+
+        :param expression:
+        :param connection:
+        :param context:
```
I'm not sure what to do to document these variables.  This method signature is the way it is to override the django method, but I don't take advantage of these variables.

jortel replied:
Might be prudent to call super() first and then decode that value.

Something like:
```
value = super(JSONField, self).from_db_value(value, expression, connection, context)
if isinstance(value, six.string_types):
    return self.to_python(value)
else:
    return value
```

pcreech replied:
Per [the docs](https://docs.djangoproject.com/en/1.8/ref/models/fields/#django.db.models.Field.from_db_value)

> For performance reasons, from_db_value is not implemented as a no-op on fields which do not require it (all Django fields). Consequently you may not call super in your definition.

bmbouter replied:
Consider these two resolutions to this docstring issue. One would be to adjust the function signature to be:

`from_db_value(self, value, *args, **kwargs):`

And then document it with:

```
:param args: unused positional arguments
:type args: list
:param kwargs: unused keyword arguments
:type kwargs: dict
```
This has the advantage of being resilient to most signature changes to the function as Django evolves. I've had this problem before with Django specifically. I think I prefer this option.

The other option is to link to the docs. If you do that I recommend pointing at latest and not just the 1.8 version since one day we will upgrade.

pcreech commented on platform/pulp/platform/fields.py:
```diff
@@ -0,0 +1,96 @@
+
+        :param connection:
+
+        :param prepared:
```
I'm not sure what to do to document these variables.  This method signature is the way it is to override the django method, but I don't take advantage of these variables.

bmbouter replied:
Similar recommendation here as above.

pcreech commented:
I've updated the PR with the JSONField implemenation

pcreech commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,231 @@
+    :cvar worker: The worker associated with this reservation
+    :type worker: models.ForeignKey
+    """
+    resource = models.TextField()
```
This needs to be unique, looks like i forgot to add it

pcreech commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,231 @@
+    """
+    resource = models.TextField()
+
+    task = models.ForeignKey("Task", unique=True)
```
Changing this to a 1-1 field, per Django's suggestion:

`pulp_platform.ReservedResource.task: (fields.W342) Setting unique=True on a ForeignKey has the same effect as using a OneToOneField.
        HINT: ForeignKey(unique=True) is usually better served by a OneToOneField.`

jortel commented on platform/pulp/platform/fields.py:
```diff
@@ -0,0 +1,96 @@
+from django.utils.translation import six, ugettext_lazy as _
+
+
+def _parse_json(json_string):
```
Instead of a separate function, perhaps you could put this code in `JSONField.to_python()` and call `self.to_python()` from `JSONField.from_db_value()`.

pcreech replied:
Good point.  The development was based off of 

https://docs.djangoproject.com/en/1.8/howto/custom-model-fields/#converting-values-to-python-objects

Which has separate code in the functions to catch different scenarios.  I originally had followed something similar, while now both functions are identical.  Should be no problem to move it into to_python, and we can always move it back out if needed.

jortel commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,231 @@
+    :type last_updated: models.DateTimeField
+
+    :cvar args: Arguments that should be passed to the apply_async function
+    :type args: models.TextField
```
JSONField

jortel commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,231 @@
+    :type args: models.TextField
+
+    :cvar kwargs: Keyword arguments that should be passed to the apply_async function
+    :type kwargs: models.TextField
```
JSONField

jortel commented:
This **LGTM** but want @bmbouter to add the tag when he's finished with his review.

I think the `JSONField` needs unit tests.

You should consider adding some tests for the models.  Mainly to provide working examples for how to interact with the json fields.

pcreech commented:
> I think the JSONField needs unit tests.
>
> You should consider adding some tests for the models. Mainly to provide working examples for how to interact with the json fields.

Working on this now, just wanted to make sure things settled down first

bmbouter commented on platform/pulp/platform/fields.py:
```diff
@@ -0,0 +1,80 @@
+"""
```
What do you think about having this in `pulp.platform.models.fields` perhaps? It's for the models layer so I figured it would be in there somewhere.

pcreech replied:
I'm not to keen on it.  Looking at other code, it seems standard to have a separate fields.py and models.py  (While we use a package instead of a module, this is the reason we import models in the __init__.py iirc)

bmbouter replied:
My comment wasn't well written. Yes fields.py and models.py should be separate. Where do Django app layouts recommend the fields.py live?

pcreech replied:
The examples I've seen put them in the same folder as models.py

bmbouter replied:
Cool I'm satisfied with this.

bmbouter commented on platform/pulp/platform/fields.py:
```diff
@@ -0,0 +1,80 @@
+
+class JSONField(models.TextField):
+    """
+    A custom django field to serialize dicts into a text field
```
s/django/Django/
s/text field/text field and vice versa/

bmbouter commented on platform/pulp/platform/fields.py:
```diff
@@ -0,0 +1,80 @@
+        Converts a value as returned by the database to a Python object
+
+        :param value: The value to convert to Python
+        :type value: Any
```
Any isn't a type but everything in Python is an object.   s/Any/object/  ?

bmbouter commented on platform/pulp/platform/fields.py:
```diff
@@ -0,0 +1,80 @@
+        :param context:
+
+        :return: A python representation of value
+        :rtype: Any
```
s/Any/object/

bmbouter commented on platform/pulp/platform/fields.py:
```diff
@@ -0,0 +1,80 @@
+        :param connection:
+        :param context:
+
+        :return: A python representation of value
```
s/python/Python/

bmbouter commented on platform/pulp/platform/fields.py:
```diff
@@ -0,0 +1,80 @@
+        """
+        if isinstance(value, six.string_types):
+            return self.to_python(value)
+
```
I generally don't put whitespace before the return statement. Pep 8 recommends using whitespace sparingly to separate logical blocks of code. It's a preference thing so this is not a required change but I wanted to mention it. It's possible someone would miss the return value since they already see one above.

bmbouter commented on platform/pulp/platform/fields.py:
```diff
@@ -0,0 +1,80 @@
+        :param value: The value to convert to Python
+        :type value: Any
+
+        :return: A python representation of value
```
s/python/Python/
I'll stop commenting on these, but similar changes throughout

bmbouter commented on platform/pulp/platform/fields.py:
```diff
@@ -0,0 +1,80 @@
+        Converts the value into the correct Python object
+
+        :param value: The value to convert to Python
+        :type value: Any
```
s/Any/object/
I'll stop commenting on these, but similar changes throughout.

bmbouter commented on platform/pulp/platform/fields.py:
```diff
@@ -0,0 +1,80 @@
+        :type value: Any
+
+        :return: A python representation of value
+        :rtype: Any
```
This should come with a `:raises ValidationError:` line documenting the exception possibility.

bmbouter commented on platform/pulp/platform/fields.py:
```diff
@@ -0,0 +1,80 @@
+        """
+        try:
+            return json.loads(value)
+        except Exception:
```
This would be better to be more restrictive. dumps() raises `ValueError` in situations where it couldn't deserialize due to a validation problem. s/Exception/ValueError/

bmbouter commented on platform/pulp/platform/fields.py:
```diff
@@ -0,0 +1,80 @@
+        try:
+            return json.loads(value)
+        except Exception:
+            raise ValidationError(_("Invalid JSON stored in database"))
```
Pulp does this pattern in a lot of places and it is difficult to debug. I think anytime we catch an exception and mutate it we should log the exception at error level. Assuming you have a logger as `_logger` you could add the line which will log the entire traceback and the bad data as well.

`_logger.exception(_('The field "value" cannot be JSON deserialized. "value" has the contents: %s') % value)`

pcreech replied:
After careful evalution, I'm thinking removing the exception handling here instead.  All this is doing is hiding information and re-raising, which (especially in this early stage) we don't want.  Especially here where we have invalid json stored in the db somehow.  My thought is to remove this, and let it bubble up to be handled by the normal mechanisms.  One example I saw only caught the ValueError to explicitly ignore it instead.

Thoughts?

bmbouter commented on platform/pulp/platform/fields.py:
```diff
@@ -0,0 +1,80 @@
+        if value is None:
+            return None
+
+        return json.dumps(value)
```
Is the function we are overriding prepared to have a `ValueError` raised or should this be caught and transformed into another exception type similar to above?

Either way this method probably needs a `:raises:` statement in the docblock.

pcreech replied:
Is it common practice to document all possible exceptions that could be thrown from dependent code that you didn't write?  I'm more familiar with having to document things you specifically throw, but not the case of all possible exceptions from third parties.

bmbouter commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,231 @@
+        (ERROR, 'Error'),
+        (CANCELED, 'Canceled')
+    )
+    type = models.TextField()
```
We can't use this name `type` because it's a reserve Python keyword.

bmbouter replied:
Going back to `task_type` would also match the docstring. That is my recommendation.

bmbouter replied:
Actually this can just be removed. We don't use this anymore and it's been deprecated already.

http://docs.pulpproject.org/dev-guide/integration/rest-api/tasks.html#task-report

pcreech replied:
Will remove, but as a point for future implementations, I shall provide @jortel 's quote on the Repository model PR about "type" variable names as opposed to using a redundant {classname}_type variable.

> Class and instance attributes are namespaced to the class (as I'm sure you know) and implicitly apply to the class or instance. The type() builtin function (not a reserved word) unfortunately exists in the global namespace. I'm opposed to redundantly qualifying it. Naming it something like repository_name does not add clarity and leads to awkward semantics like: repository.repository_type.
https://docs.python.org/2/library/functions.html#type

bmbouter commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,231 @@
+        (RUNNING, 'Running'),
+        (SUSPENDED, 'Suspended'),
+        (FINISHED, 'Finished'),
+        (ERROR, 'Error'),
```
I'm hoping `Error` can be `Errored` so match the tense of all other states. I'm making this change on my progress reporting models too.

bmbouter commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,231 @@
+        (ACCEPTED, 'Accepted'),
+        (RUNNING, 'Running'),
+        (SUSPENDED, 'Suspended'),
+        (FINISHED, 'Finished'),
```
In the progress reporting models I'm changing `Finished` to `Completed` so that it conveys the idea that it finished without error. This is something I'm thinking about and not a required change.

bmbouter commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,231 @@
+    RUNNING = 'running'
+    SUSPENDED = 'suspended'
+    FINISHED = 'finished'
+    ERROR = 'error'
```
I'm thinking about s/error/errored/ to match the tense of the others. I'm making this change in the progress reporting models too.

bmbouter commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,231 @@
+    ACCEPTED = 'accepted'
+    RUNNING = 'running'
+    SUSPENDED = 'suspended'
+    FINISHED = 'finished'
```
s/FINISHED/COMPLETED/  ?
s/finished/completed/  ?

bmbouter commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,231 @@
+    :cvar task_type: The type of task this is
+    :type task_type: models.TextField
+
+    :cvar task_group: The group this task belongs to
```
s/task_group/group/ to match the code.

bmbouter commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,231 @@
+    :type task_type: models.TextField
+
+    :cvar task_group: The group this task belongs to
+    :type task_group: models.UUIDField
```
s/task_group/group/ to match the code.

bmbouter commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,231 @@
+    error = JSONField()
+    result = JSONField()
+
+    parent = models.ForeignKey("Task", null=True)
```
This makes sense for the data model, but I want to point something out. In terms of the purpose of this field to present this data to the user the parent has to name its children. Leaving it as-is is right for the data layer, but we need to have `spawned_tasks` or some similar name be returned when a Task is serialized through the API. We should capture that somehow or somewhere. I'm not sure what to do about that.

pcreech replied:
From the look of the docs, this relationship can be [followed backwards](https://docs.djangoproject.com/en/1.8/topics/db/queries/#following-relationships-backward) i.e. the parent will have a task_set value that you can use to get the children (and somewhere else in the code can be used to convert this into something useful for the api)  We should be able to rename it via related_name to something more intuitive

bmbouter replied:
Oh yes I remember related_name. Is this something we would define on the class attribute L#141 or at the API representation level?

bmbouter replied:
Ignore this comment, you already fixed it.

bmbouter replied:
The `spawned_tasks` needs to go on `parent` not on `worker`.

bmbouter commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,231 @@
+    result = JSONField()
+
+    parent = models.ForeignKey("Task", null=True)
+    worker = models.ForeignKey("Worker")
```
There is a time when a Task is in the DB, but Worker assignment is not yet known. This needs to allow null I think.

bmbouter commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,231 @@
+    enabled = models.BooleanField(default=True)
+    resource = models.TextField(null=True)
+
+    consecutive_failures = models.IntegerField()
```
The feature that requires this is being removed from 3.0 so this can be removed from the data layer.

bmbouter commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,231 @@
+    resource = models.TextField(null=True)
+
+    consecutive_failures = models.IntegerField()
+    failure_threshold = models.IntegerField()
```
The feature that requires this is being removed from 3.0 so this can be removed.

bmbouter commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,231 @@
+    consecutive_failures = models.IntegerField()
+    failure_threshold = models.IntegerField()
+
+    iso_schedule = models.TextField()
```
This probably isn't a data layer concern except that I was hoping for more validation of this field.

For 3.0 we should contemplate if we want to continue to use ISO8601. I haven't met a person yet who can correctly form their intended schedule on the first try after reading the ISO8601 docs. Something like what cron uses would be so much better.

What do you think? Maybe we should send this to pulp-dev mailing list for comment?

pcreech replied:
I'm a fan for pulp-dev'ing this convo.  ISO date formats have always seemed, to me, more of a way of 'serializing/deserializing' date data for machine consumption than allowing a human-friendly way of providing this information.

IMO, We could easily provide a more user-friendly scheduling system to work with for 3.0

bmbouter commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,231 @@
+    failure_threshold = models.IntegerField()
+
+    iso_schedule = models.TextField()
+    schedule = models.TextField(null=True)
```
This says it's stored as a pickle field but we're going to store it as a `TextField`. This should work but I wanted to mention it. Maybe it will need some serialization/deserialization. We can also handle this when we hook code up to the models then since it probably will show a tangible version of the problem at that time.

pcreech replied:
I'll leave it as is for now, but we could easily turn in into a JSONField as well if needed.

bmbouter commented:
Here are two things that we should talk about how to handle:

1) In 3.0 we want to maintain the AMQP based signaling on each save of a Task. That is done in 2.y [here](https://github.com/bmbouter/pulp/blob/6868481c1a0eeab06d1f18330e7fd7dc533d8007/server/pulp/server/db/model/__init__.py#L567-L567). That would likely be a overridden save() method which calls super() and then sends the message. We could leave the send_message implementation blank for now and file an issue to wire that up later.

2) The reaper Mixin model which reaps Tasks along with a bunch of other model types. This is likely it's own separate problem so I would be satisfied if an issue was created to make sure this works in 3.0 with our new models.

bmbouter commented:
@pcreech I'm done w/ my review. I left a bunch of comments. Functionally this is a great start and will likely be easy to make work with the existing tasking system code. Want to fix some of the things and answer some of the questions and ping me with quesitons or to do another pass. I expect the next pass to be the last one. Thanks for this great work!

pcreech commented:
> 1) In 3.0 we want to maintain the AMQP based signaling on each save of a Task. That is done in 2.y here. That would likely be a overridden save() method which calls super() and then sends the message. We could leave the send_message implementation blank for now and file an issue to wire that up later.

In response to this statement specifically, we should avoid putting any code not related directly to managing the model in the model.

Django appears to have a robust system of signal handlers we can wire functions into to handle this situation.  Most importantly, in this case, is the post_save signal outlined [here](https://docs.djangoproject.com/en/1.8/ref/signals/#post-save)

When the person handling the signal wire-up does that task, they can take advantage of this to handle it.

pcreech commented:
Updated.  I tweaked some field defs as well for things like nullables or to keep defs in line with each other (i.e. strings instead of types for related fields)

bmbouter commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,218 @@
+
+    first_run = models.DateTimeField()
+
+    remaining_runs = models.IntegerField(null=True)
```
This field is also able to be deleted and won't be used in 3.0 I believe. As far as I know we don't meaningfully rely on it even in 2.y. The only [example in the docs](http://docs.pulpproject.org/dev-guide/conventions/scheduled.html?highlight=remaining_runs) shows it as null.

We could delete it now and if we need it we can bring it back.

bmbouter commented on platform/pulp/platform/models/task.py:
```diff
@@ -0,0 +1,218 @@
+    iso_schedule = models.TextField()
+    schedule = models.TextField(null=True)
+
+    first_run = models.DateTimeField()
```
Can this be grouped in with the `last_run` field since they go together? Also the docstrings could move similarly.

bmbouter commented:
I left a few more comments. You have a great way of not missing any of them so I'm going to go ahead and add the LGTM flag now. @pcreech great great work. Thank you so much! This is going to be awesome. :cat2: :fireworks: :grapes:
